src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -1715,15 +1717,50 @@ NOINLINE intptr_t* Thaw<ConfigT>::thaw_fast(stackChunkOop chunk) {
 #endif
 
 #ifdef ASSERT
-  set_anchor(_thread, stack_sp);
+  set_anchor(_thread, rs.sp());
   log_frames(_thread);
   if (LoomDeoptAfterThaw) {
     do_deopt_after_thaw(_thread);
   }
   clear_anchor(_thread);
 #endif
 
-  return stack_sp;
+  return rs.sp();
+}
+
+inline void ThawBase::clear_chunk(stackChunkOop chunk) {
+  chunk->set_sp(chunk->stack_size());
+  chunk->set_argsize(0);
+  chunk->set_max_thawing_size(0);
+}
+
+ int ThawBase::remove_top_compiled_frame_from_chunk(stackChunkOop chunk, int &argsize) {
+  bool empty = false;
+  StackChunkFrameStream<ChunkFrames::CompiledOnly> f(chunk);
+  DEBUG_ONLY(intptr_t* const chunk_sp = chunk->start_address() + chunk->sp();)
+  assert(chunk_sp == f.sp(), "");
+  assert(chunk_sp == f.unextended_sp(), "");
+
+  const int frame_size = f.cb()->frame_size();
+  argsize = f.stack_argsize();
+
+  f.next(SmallRegisterMap::instance, true /* stop */);
+  empty = f.is_done();
+  assert(!empty || argsize == chunk->argsize(), "");
+
+  if (empty) {
+    clear_chunk(chunk);
+  } else {
+    chunk->set_sp(chunk->sp() + frame_size);
+    chunk->set_max_thawing_size(chunk->max_thawing_size() - frame_size);
+    // We set chunk->pc to the return pc into the next frame
+    chunk->set_pc(f.pc());
+    assert(f.pc() == *(address*)(chunk_sp + frame_size - frame::sender_sp_ret_address_offset()), "unexpected pc");
+  }
+  assert(empty == chunk->is_empty(), "");
+  // returns the size required to store the frame on stack, and because it is a
+  // compiled frame, it must include a copy of the arguments passed by the caller
+  return frame_size + argsize;
```

Comment:
small nit, can you put these functions above the ones that call them in the file?

Reply:
Sure. Methods re-ordered: clear_chunk(),  remove_top_compiled_frame_from_chunk(), copy_from_chunk() and patch_return().